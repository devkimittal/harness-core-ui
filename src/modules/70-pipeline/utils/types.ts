/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { StringNGVariable, NumberNGVariable, SecretNGVariable, PipelineInfoConfig } from 'services/cd-ng'
import type { NodeRunInfo, EntityGitDetails, EntityValidityDetails, InputSetResponse } from 'services/pipeline-ng'

export type AllNGVariables = StringNGVariable | NumberNGVariable | SecretNGVariable

export interface ExecutionPageQueryParams {
  view?: 'log' | 'graph'
  stage?: string
  step?: string
  retryStep?: string
}

export interface ExpressionBlock {
  expression?: string
  expressionValue?: string
}

export interface ConditionalExecutionNodeRunInfo extends NodeRunInfo {
  expressions?: ExpressionBlock[]
}

export type StatusType = 'SUCCESS' | 'FAILURE' | 'ERROR'

export type CreateUpdateInputSetsReturnType = Promise<{
  status?: StatusType
  nextCallback: () => void
}>

export interface InputSetType {
  name: string
  tags?: { [key: string]: string }
  identifier: string
  description: string
  orgIdentifier: string
  projectIdentifier: string
  pipeline: PipelineInfoConfig
  gitDetails: EntityGitDetails
  entityValidityDetails: EntityValidityDetails
}

export interface InputSetDTO extends Omit<InputSetResponse, 'identifier' | 'pipeline'> {
  pipeline?: PipelineInfoConfig
  identifier?: string
  repo?: string
  branch?: string
}

export interface SaveInputSetDTO {
  inputSet: InputSetDTO
}
