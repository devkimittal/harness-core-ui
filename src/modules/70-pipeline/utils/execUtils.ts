import type { PipelineGraphState } from '@pipeline/components/AbstractNode/types'
import type { ExecutionPipelineNode } from '@pipeline/components/ExecutionStageDiagram/ExecutionPipelineModel'
import type { ExecutionGraph, ExecutionNode } from 'services/pipeline-ng'
import {
  StepGroupRollbackIdentifier,
  NodeType,
  RollbackContainerCss,
  getIconDataBasedOnType,
  LITE_ENGINE_TASK,
  getExecutionPipelineNodeType,
  processLiteEngineTask
} from './executionUtils'
import type { ExecutionStatus } from './statusHelpers'

export const processNodeDataV1 = (
  children: string[],
  nodeMap: ExecutionGraph['nodeMap'],
  nodeAdjacencyListMap: ExecutionGraph['nodeAdjacencyListMap'],
  rootNodes: PipelineGraphState[]
): any[] => {
  const items: any[] = []
  children?.forEach(item => {
    const nodeData = nodeMap?.[item]
    const isRollback = nodeData?.name?.endsWith(StepGroupRollbackIdentifier) ?? false
    if (nodeData?.stepType === NodeType.FORK) {
      const [parentNode, ...childNodes] = nodeAdjacencyListMap?.[item].children as any[]

      const iconData = getIconDataBasedOnType(parentNode)
      items.push({
        id: item as string,
        identifier: item as string,
        name: parentNode?.name as string,
        icon: iconData.icon,
        type: 'StepGroup',
        nodeType: getExecutionPipelineNodeType(parentNode?.stepType),
        data: {
          name: parentNode.name || /* istanbul ignore next */ '',
          identifier: item,
          data: parentNode,
          containerCss: {
            ...(isRollback ? RollbackContainerCss : {})
          },
          status: parentNode.status as ExecutionStatus,
          isOpen: true,
          skipCondition: parentNode.skipInfo?.evaluatedCondition ? parentNode.skipInfo.skipCondition : undefined,
          when: parentNode.nodeRunInfo,
          ...iconData
        },
        children: processNodeDataV1(
          childNodes || /* istanbul ignore next */ [],
          nodeMap,
          nodeAdjacencyListMap,
          rootNodes
        )
      })
    } else if (
      nodeData?.stepType === NodeType.STEP_GROUP ||
      nodeData?.stepType === NodeType.NG_SECTION ||
      (nodeData && isRollback)
    ) {
      const iconData = getIconDataBasedOnType(nodeData)
      items.push({
        id: item as string,
        identifier: item as string,
        name: nodeData?.name as string,
        icon: iconData.icon,
        type: 'StepGroup',
        nodeType: getExecutionPipelineNodeType(nodeData?.stepType),
        data: {
          name: nodeData.name || /* istanbul ignore next */ '',
          identifier: item,
          data: nodeData,
          containerCss: {
            ...(isRollback ? RollbackContainerCss : {})
          },
          status: nodeData.status as ExecutionStatus,
          isOpen: true,
          skipCondition: nodeData.skipInfo?.evaluatedCondition ? nodeData.skipInfo.skipCondition : undefined,
          when: nodeData.nodeRunInfo,
          ...getIconDataBasedOnType(nodeData)
        },
        children: processNodeDataV1(
          nodeAdjacencyListMap?.[item].children || /* istanbul ignore next */ [],
          nodeMap,
          nodeAdjacencyListMap,
          rootNodes
        )
      } as PipelineGraphState)
    } else {
      if (nodeData?.stepType === LITE_ENGINE_TASK) {
        const parentNodeId =
          Object.entries(nodeAdjacencyListMap || {}).find(([_, val]) => {
            return (val?.children?.indexOf(nodeData.uuid!) ?? -1) >= 0
          })?.[0] || ''
        processLiteEngineTask(nodeData, rootNodes, nodeMap?.[parentNodeId])
      } else {
        const iconData = getIconDataBasedOnType(nodeData)
        items.push({
          id: item as string,
          identifier: item as string,
          name: nodeData?.name as string,
          icon: iconData.icon,
          type: nodeData?.stepType,
          nodeType: getExecutionPipelineNodeType(nodeData?.stepType),
          data: {
            name: nodeData?.name || /* istanbul ignore next */ '',
            ...iconData,
            identifier: item,
            skipCondition: nodeData?.skipInfo?.evaluatedCondition ? nodeData?.skipInfo.skipCondition : undefined,
            when: nodeData?.nodeRunInfo,
            status: nodeData?.status as ExecutionStatus,
            type: getExecutionPipelineNodeType(nodeData?.stepType),
            data: nodeData
          }
        } as PipelineGraphState)
      }
    }
    const nextIds = nodeAdjacencyListMap?.[item].nextIds || /* istanbul ignore next */ []
    nextIds.forEach(id => {
      const nodeDataNext = nodeMap?.[id]
      const isRollbackNext = nodeDataNext?.name?.endsWith(StepGroupRollbackIdentifier) ?? false
      if (nodeDataNext?.stepType === NodeType.FORK) {
        // items.push({
        //   parallel: processNodeDataV1(
        //     nodeAdjacencyListMap?.[id].children || /* istanbul ignore next */ [],
        //     nodeMap,
        //     nodeAdjacencyListMap,
        //     rootNodes
        //   )
        // })

        const [parentNode, ...childNodes] = nodeAdjacencyListMap?.[id].children as any[]

        const iconData = getIconDataBasedOnType(parentNode)
        items.push({
          id: item as string,
          identifier: item as string,
          name: parentNode?.name as string,
          icon: iconData.icon,
          type: 'StepGroup',
          nodeType: getExecutionPipelineNodeType(parentNode?.stepType),
          data: {
            name: parentNode.name || /* istanbul ignore next */ '',
            identifier: item,
            data: parentNode,
            containerCss: {
              ...(isRollback ? RollbackContainerCss : {})
            },
            status: parentNode.status as ExecutionStatus,
            isOpen: true,
            skipCondition: parentNode.skipInfo?.evaluatedCondition ? parentNode.skipInfo.skipCondition : undefined,
            when: parentNode.nodeRunInfo,
            ...iconData
          },
          children: processNodeDataV1(
            childNodes || /* istanbul ignore next */ [],
            nodeMap,
            nodeAdjacencyListMap,
            rootNodes
          )
        })
      } else if (nodeDataNext?.stepType === NodeType.STEP_GROUP || (isRollbackNext && nodeDataNext)) {
        // type: 'StepGroup',
        // nodeType: 'StepGroup',

        const iconData = getIconDataBasedOnType(nodeDataNext)
        items.push({
          id: item as string,
          identifier: item as string,
          name: nodeData?.name as string,
          icon: iconData.icon,
          type: 'StepGroup',
          nodeType: 'StepGroup',
          data: {
            name: nodeDataNext.name || /* istanbul ignore next */ '',
            identifier: id,
            data: nodeDataNext,
            containerCss: {
              ...(isRollbackNext ? RollbackContainerCss : {})
            },
            skipCondition: nodeDataNext.skipInfo?.evaluatedCondition ? nodeDataNext.skipInfo.skipCondition : undefined,
            when: nodeDataNext.nodeRunInfo,
            status: nodeDataNext.status as ExecutionStatus,
            isOpen: true,
            ...iconData
          },
          children: processNodeDataV1(
            nodeAdjacencyListMap?.[id].children || /* istanbul ignore next */ [],
            nodeMap,
            nodeAdjacencyListMap,
            rootNodes
          )
        })
      } else {
        const iconData = getIconDataBasedOnType(nodeDataNext)
        items.push({
          id: id as string,
          identifier: id as string,
          name: nodeDataNext?.name as string,
          icon: iconData.icon,
          type: nodeDataNext?.stepType,
          nodeType: getExecutionPipelineNodeType(nodeDataNext?.stepType),
          data: {
            name: nodeDataNext?.name || /* istanbul ignore next */ '',
            ...iconData,
            identifier: id,
            skipCondition: nodeDataNext?.skipInfo?.evaluatedCondition ? nodeDataNext.skipInfo.skipCondition : undefined,
            when: nodeDataNext?.nodeRunInfo,
            status: nodeDataNext?.status as ExecutionStatus,
            type: getExecutionPipelineNodeType(nodeDataNext?.stepType),
            data: nodeDataNext
          },
          children: []
        } as PipelineGraphState)
      }
      const nextLevels = nodeAdjacencyListMap?.[id].nextIds
      if (nextLevels) {
        items.push(...processNodeDataV1(nextLevels, nodeMap, nodeAdjacencyListMap, rootNodes))
      }
    })
  })
  return items
}
