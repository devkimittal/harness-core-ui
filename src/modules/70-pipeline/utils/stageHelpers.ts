/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo, get, isEmpty } from 'lodash-es'
import { getMultiTypeFromValue, MultiTypeInputType } from '@wings-software/uicore'
import type { GraphLayoutNode, PipelineExecutionSummary } from 'services/pipeline-ng'
import type { StringKeys } from 'framework/strings'
import type { GetExecutionStrategyYamlQueryParams, PipelineInfoConfig, StageElementConfig } from 'services/cd-ng'
import { ManifestDataType } from '@pipeline/components/ManifestSelection/Manifesthelper'
import type { ManifestTypes } from '@pipeline/components/ManifestSelection/ManifestInterface'
import type { InputSetDTO } from './types'
import type { DeploymentStageElementConfig, PipelineStageWrapper, StageElementWrapper } from './pipelineTypes'

export enum StageType {
  DEPLOY = 'Deployment',
  BUILD = 'CI',
  FEATURE = 'FeatureFlag',
  PIPELINE = 'Pipeline',
  APPROVAL = 'Approval',
  CUSTOM = 'Custom',
  Template = 'Template',
  SECURITY = 'Security'
}

export enum ServiceDeploymentType {
  Kubernetes = 'Kubernetes',
  NativeHelm = 'NativeHelm',
  amazonEcs = 'amazonEcs',
  amazonAmi = 'amazonAmi',
  awsCodeDeploy = 'awsCodeDeploy',
  winrm = 'winrm',
  awsLambda = 'awsLambda',
  pcf = 'pcf',
  ssh = 'ssh',
  ServerlessAwsLambda = 'ServerlessAwsLambda',
  ServerlessAzureFunction = 'ServerlessAzureFunctions',
  ServerlessGoogleFunctions = 'ServerlessGoogleFunctions',
  AmazonSAM = 'AwsSAM',
  AzureFunctions = 'AzureFunctions'
}

export const changeEmptyValuesToRunTimeInput = (inputset: any, propertyKey: string): InputSetDTO => {
  if (inputset) {
    Object.keys(inputset).forEach(key => {
      if (typeof inputset[key] === 'object') {
        changeEmptyValuesToRunTimeInput(inputset[key], key)
      } else if (inputset[key] === '' && ['tags'].indexOf(propertyKey) === -1) {
        inputset[key] = '<+input>'
      }
    })
  }
  return inputset
}

export function isCDStage(node?: GraphLayoutNode): boolean {
  return node?.nodeType === StageType.DEPLOY || node?.module === 'cd' || !isEmpty(node?.moduleInfo?.cd)
}

export function isCIStage(node?: GraphLayoutNode): boolean {
  return node?.nodeType === StageType.BUILD || node?.module === 'ci' || !isEmpty(node?.moduleInfo?.ci)
}

export function hasCDStage(pipelineExecution?: PipelineExecutionSummary): boolean {
  return pipelineExecution?.modules?.includes('cd') || !isEmpty(pipelineExecution?.moduleInfo?.cd)
}

export function hasCIStage(pipelineExecution?: PipelineExecutionSummary): boolean {
  return pipelineExecution?.modules?.includes('ci') || !isEmpty(pipelineExecution?.moduleInfo?.ci)
}

export const getHelperTextString = (
  invalidFields: string[],
  getString: (key: StringKeys) => string,
  isServerlessDeploymentTypeSelected = false
): string => {
  return `${invalidFields.length > 1 ? invalidFields.join(', ') : invalidFields[0]} ${
    invalidFields.length > 1 ? ' are ' : ' is '
  } ${
    isServerlessDeploymentTypeSelected
      ? getString('pipeline.artifactPathDependencyRequired')
      : getString('pipeline.tagDependencyRequired')
  }`
}

export const getHelpeTextForTags = (
  fields: {
    imagePath?: string
    artifactPath?: string
    region?: string
    connectorRef: string
    registryHostname?: string
    repository?: string
    repositoryPort?: number
    artifactDirectory?: string
  },
  getString: (key: StringKeys) => string,
  isServerlessDeploymentTypeSelected = false
): string => {
  const {
    connectorRef,
    region,
    imagePath,
    artifactPath,
    registryHostname,
    repository,
    repositoryPort,
    artifactDirectory
  } = fields
  const invalidFields: string[] = []
  if (!connectorRef || getMultiTypeFromValue(connectorRef) === MultiTypeInputType.RUNTIME) {
    invalidFields.push(getString('connector'))
  }
  if (region !== undefined && (!region || getMultiTypeFromValue(region) === MultiTypeInputType.RUNTIME)) {
    invalidFields.push(getString('regionLabel'))
  }
  if (
    registryHostname !== undefined &&
    (!registryHostname || getMultiTypeFromValue(registryHostname) === MultiTypeInputType.RUNTIME)
  ) {
    invalidFields.push(getString('connectors.GCR.registryHostname'))
  }
  if (
    !isServerlessDeploymentTypeSelected &&
    (!imagePath || getMultiTypeFromValue(imagePath) === MultiTypeInputType.RUNTIME)
  ) {
    invalidFields.push(getString('pipeline.imagePathLabel'))
  }
  if (
    !isServerlessDeploymentTypeSelected &&
    (!imagePath || getMultiTypeFromValue(artifactPath) === MultiTypeInputType.RUNTIME)
  ) {
    invalidFields.push(getString('pipeline.artifactPathLabel'))
  }
  if (repository !== undefined && (!repository || getMultiTypeFromValue(repository) === MultiTypeInputType.RUNTIME)) {
    invalidFields.push(getString('repository'))
  }
  if (
    repositoryPort !== undefined &&
    (!repositoryPort || getMultiTypeFromValue(repositoryPort) === MultiTypeInputType.RUNTIME)
  ) {
    invalidFields.push(getString('pipeline.artifactsSelection.repositoryPort'))
  }
  if (
    isServerlessDeploymentTypeSelected &&
    (!artifactDirectory || getMultiTypeFromValue(artifactDirectory) === MultiTypeInputType.RUNTIME)
  ) {
    invalidFields.push(getString('pipeline.artifactsSelection.artifactDirectory'))
  }

  const helpText = getHelperTextString(invalidFields, getString, isServerlessDeploymentTypeSelected)

  return invalidFields.length > 0 ? helpText : ''
}

export const isServerlessDeploymentType = (deploymentType: string): boolean => {
  return (
    deploymentType === ServiceDeploymentType.ServerlessAwsLambda ||
    deploymentType === ServiceDeploymentType.ServerlessAzureFunction ||
    deploymentType === ServiceDeploymentType.ServerlessGoogleFunctions ||
    deploymentType === ServiceDeploymentType.AmazonSAM ||
    deploymentType === ServiceDeploymentType.AzureFunctions
  )
}

export const isServerlessManifestType = (selectedManifest: ManifestTypes | null): boolean => {
  return selectedManifest === ManifestDataType.ServerlessAwsLambda
}

export const getSelectedDeploymentType = (
  stage: StageElementWrapper<DeploymentStageElementConfig> | undefined,
  getStageFromPipeline: <T extends StageElementConfig = StageElementConfig>(
    stageId: string,
    pipeline?: PipelineInfoConfig | undefined
  ) => PipelineStageWrapper<T>,
  isPropagating = false
): GetExecutionStrategyYamlQueryParams['serviceDefinitionType'] => {
  if (isPropagating) {
    const parentStageId = get(stage, 'stage.spec.serviceConfig.useFromStage.stage', null)
    const parentStage = getStageFromPipeline<DeploymentStageElementConfig>(defaultTo(parentStageId, ''))
    return get(parentStage, 'stage.stage.spec.serviceConfig.serviceDefinition.type', null)
  }
  return get(stage, 'stage.spec.serviceConfig.serviceDefinition.type', null)
}
