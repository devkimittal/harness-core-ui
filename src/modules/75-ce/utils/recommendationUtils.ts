/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export const addBufferToValue = (value: number, bufferPercentage: number): number =>
  +(((100 + bufferPercentage) / 100) * value).toFixed(2)

export const calculateSavingsPercentage = (savings: number, totalCost: number): string =>
  `(${Math.floor((savings / totalCost) * 100)}%)`

export const isResourceConsistent = (sumCpu: number, sumMemory: number, maxCpu: number, maxMemory: number): boolean => {
  const isInconsistent = Math.round(sumCpu) < Math.round(maxCpu) || Math.round(sumMemory) < Math.round(maxMemory)

  const anyZero =
    Math.round(sumCpu) == 0 || Math.round(sumMemory) == 0 || Math.round(maxCpu) == 0 || Math.round(maxMemory) == 0

  return !isInconsistent || !anyZero
}

export const calculateNodes = (
  sumCpu: number,
  sumMemory: number,
  maxCpu: number,
  maxMemory: number,
  minNodes: number
): { maximumNodes: number; minimumNodes: number } => {
  let minimumNodes = minNodes || 3

  let maximumNodes = Math.min(Math.floor(sumCpu / maxCpu), Math.floor(sumMemory / maxMemory))
  maximumNodes = Math.max(maximumNodes, 1)

  if (maximumNodes < minNodes) {
    minimumNodes = maximumNodes
  }

  return { maximumNodes, minimumNodes }
}
