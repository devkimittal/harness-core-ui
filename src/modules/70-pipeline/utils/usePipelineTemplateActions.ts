/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo, isEqual } from 'lodash-es'
import { parse } from 'yaml'
import produce from 'immer'
import { useCallback } from 'react'
import type { PipelineInfoConfig } from 'services/cd-ng'
import type { TemplateSummaryResponse } from 'services/template-ng'
import { usePipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import { useTemplateSelector } from '@pipeline/utils/useTemplateSelector'
import { createTemplate } from '@pipeline/utils/templateUtils'
import { getIdentifierFromValue } from '@common/components/EntityReference/EntityReference'

interface TemplateActionsReturnType {
  onUseTemplate: (template: TemplateSummaryResponse, isCopied?: boolean) => Promise<void>
  onRemoveTemplate: () => Promise<void>
  onOpenTemplateSelector: () => void
}

export function usePipelineTemplateActions(): TemplateActionsReturnType {
  const {
    state: { pipeline, templateTypes },
    setTemplateTypes,
    updatePipeline
  } = usePipelineContext()
  const { openTemplateSelector, closeTemplateSelector } = useTemplateSelector()

  const onUseTemplate = useCallback(
    async (template: TemplateSummaryResponse, isCopied = false) => {
      closeTemplateSelector()
      if (
        !isCopied &&
        isEqual(pipeline?.template?.templateRef, template.identifier) &&
        isEqual(pipeline?.template?.versionLabel, template.versionLabel)
      ) {
        return
      }
      const processNode = isCopied
        ? produce(defaultTo(parse(template?.yaml || '')?.template.spec, {}) as PipelineInfoConfig, draft => {
            draft.name = defaultTo(pipeline?.name, '')
            draft.identifier = defaultTo(pipeline?.identifier, '')
          })
        : createTemplate(pipeline, template)
      await updatePipeline(processNode)
      if (!isCopied && template?.identifier && template?.childType) {
        templateTypes[template.identifier] = template.childType
        setTemplateTypes(templateTypes)
      }
    },
    [closeTemplateSelector, pipeline, updatePipeline]
  )

  const onOpenTemplateSelector = useCallback(() => {
    openTemplateSelector({
      templateType: 'Pipeline',
      selectedTemplateRef: getIdentifierFromValue(defaultTo(pipeline.template?.templateRef, '')),
      selectedVersionLabel: pipeline.template?.versionLabel,
      onUseTemplate
    })
  }, [pipeline.template, openTemplateSelector, onUseTemplate])

  const onRemoveTemplate = useCallback(async () => {
    const node = pipeline
    const processNode = produce({} as PipelineInfoConfig, draft => {
      draft.name = defaultTo(node?.name, '')
      draft.identifier = defaultTo(node?.identifier, '')
    })
    await updatePipeline(processNode)
  }, [pipeline, updatePipeline])

  return { onUseTemplate, onRemoveTemplate, onOpenTemplateSelector }
}
