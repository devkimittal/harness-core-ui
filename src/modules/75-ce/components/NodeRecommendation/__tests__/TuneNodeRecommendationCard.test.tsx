/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import TuneRecommendationCard from '../TuneNodeRecommendationCard'

const mockState = {
  minCpu: 44,
  minMem: 100,
  sumCpu: 30,
  sumMem: 120,
  maxNodes: 12,
  minNodes: 10,
  includeTypes: [],
  includeSeries: [],
  excludeTypes: [],
  excludeSeries: []
}

describe('test cases for tune recommendation card', () => {
  test('should be able to render tune recommendation card', async () => {
    const { container } = render(
      <TestWrapper>
        <TuneRecommendationCard buffer={0} dispatch={jest.fn()} setBuffer={jest.fn()} state={mockState} />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })
})
