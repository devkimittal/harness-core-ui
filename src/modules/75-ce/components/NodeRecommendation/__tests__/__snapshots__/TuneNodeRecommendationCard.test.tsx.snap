// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test cases for recommendation tabs should be able to render recommendation tabs 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main preferences"
  >
    <div
      class="Layout--vertical Layout--layout-spacing-medium StyledProps--main"
    >
      <div
        class="StyledProps--font StyledProps--main"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-small-semi"
        >
          ce.nodeRecommendation.prefResourceNeeds
        </p>
        <div
          class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between StyledProps--margin-top-small"
        >
          <div
            class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
          >
            <div
              class="StyledProps--font StyledProps--main"
            >
              <div
                class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
              >
                <div
                  class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
                >
                  <span
                    class="StyledProps--font StyledProps--main StyledProps--padding-right-small StyledProps--inline StyledProps--color StyledProps--color-grey500 StyledProps--font-variation-small-semi"
                  >
                    ce.nodeRecommendation.cpus
                  </span>
                  <div
                    class="TextInput--main input"
                  >
                    <div
                      class="bp3-input-group"
                    >
                      <input
                        class="bp3-input"
                        type="text"
                        value="30"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
                >
                  <span
                    class="StyledProps--font StyledProps--main StyledProps--padding-right-small StyledProps--inline StyledProps--color StyledProps--color-grey500 StyledProps--font-variation-small-semi"
                  >
                    ce.nodeRecommendation.mem
                  </span>
                  <div
                    class="TextInput--main input"
                  >
                    <div
                      class="bp3-input-group"
                    >
                      <input
                        class="bp3-input"
                        type="text"
                        value="120"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-center"
            >
              <div
                class="StyledProps--font StyledProps--main"
                style="border-width: 1px; border-style: solid; border-left: 0; width: 14px; height: 60px;"
              />
            </div>
            <div
              class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-center StyledProps--margin-right-small"
            >
              <span
                class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--color StyledProps--color-grey700"
                icon="plus"
              >
                <svg
                  data-icon="plus"
                  height="12"
                  viewBox="0 0 16 16"
                  width="12"
                >
                  <desc>
                    plus
                  </desc>
                  <path
                    d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            class="StyledProps--font StyledProps--main"
            style="width: 60%;"
          >
            <div
              class="Layout--vertical StyledProps--main"
            >
              <div
                class="Layout--horizontal StyledProps--main"
                style="justify-content: space-between;"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--font-variation-small-semi"
                  style="display: flex; align-items: center;"
                >
                  ce.nodeRecommendation.buffer
                  <span
                    class="bp3-icon StyledProps--main Text--icon StyledProps--padding-left-xsmall"
                    data-icon="info"
                  >
                    <svg
                      height="12"
                      viewBox="0 0 12 12"
                      width="12"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M0 6a6 6 0 1012 0A6 6 0 000 6zm.75 0a5.25 5.25 0 1010.5 0A5.25 5.25 0 00.75 6zm4.206 3.104v.646h2.682v-.646l-.836-.144V4.203H4.956v.65l.836.144V8.96l-.836.144zm1.846-7.352v1.03h-1.01v-1.03h1.01z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </p>
                <p
                  class="StyledProps--font StyledProps--main StyledProps--font-variation-small-semi StyledProps--color StyledProps--color-grey400"
                >
                  0%
                </p>
              </div>
              <div
                class="bp3-slider bp3-slider-unlabeled bufferSlider"
              >
                <div
                  class="bp3-slider-track"
                >
                  <div
                    class="bp3-slider-progress bp3-intent-primary"
                    style="left: 0.00%; right: 100.00%; top: 0px;"
                  />
                  <div
                    class="bp3-slider-progress"
                    style="left: 0.00%; right: 100.00%; top: 0px;"
                  />
                  <div
                    class="bp3-slider-progress"
                    style="left: 0.00%; right: 0.00%; top: 0px;"
                  />
                </div>
                <div
                  class="bp3-slider-axis"
                />
                <span
                  class="bp3-slider-handle"
                  style="left: calc(0.00% - 0px);"
                  tabindex="0"
                />
              </div>
              <div
                class="StyledProps--font StyledProps--main"
              >
                <span
                  class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--font-variation-small"
                >
                  delegate.delegateCPU: 30 vCPU 
                </span>
                <span
                  class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--font-variation-small"
                >
                  ce.nodeRecommendation.ram: 120 GiB
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main StyledProps--padding-top-medium"
        >
          <div
            class="StyledProps--font StyledProps--main"
          >
            <div
              class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
            >
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-variation-small-semi"
              >
                ce.nodeRecommendation.largestWorkloadReq
              </p>
              <div
                class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
                style="width: 75%;"
              >
                <span
                  class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--color StyledProps--color-grey500 StyledProps--font-variation-small-semi"
                >
                  ce.nodeRecommendation.cpus
                </span>
                <div
                  class="TextInput--main input"
                >
                  <div
                    class="bp3-input-group"
                  >
                    <input
                      class="bp3-input"
                      type="text"
                      value="44"
                    />
                  </div>
                </div>
              </div>
              <div
                class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
                style="width: 75%;"
              >
                <span
                  class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--color StyledProps--color-grey500 StyledProps--font-variation-small-semi"
                >
                  ce.nodeRecommendation.mem
                </span>
                <div
                  class="TextInput--main input"
                >
                  <div
                    class="bp3-input-group"
                  >
                    <input
                      class="bp3-input"
                      type="text"
                      value="100"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="StyledProps--font StyledProps--main StyledProps--padding-left-large"
          >
            <div
              class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
            >
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-variation-small-semi"
              >
                ce.nodeRecommendation.prefMinNodeCount
              </p>
              <div
                class="Layout--horizontal Layout--layout-spacing-small StyledProps--main minNodeContainer"
              >
                <button
                  class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--withLeftIcon Button--variation Button--variation-secondary"
                  type="button"
                >
                  <span
                    class="bp3-icon bp3-icon-minus StyledProps--main"
                    icon="minus"
                  >
                    <svg
                      data-icon="minus"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <desc>
                        minus
                      </desc>
                      <path
                        d="M13 7H3c-.55 0-1 .45-1 1s.45 1 1 1h10c.55 0 1-.45 1-1s-.45-1-1-1z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </button>
                <div
                  class="TextInput--main minNodeInput"
                >
                  <div
                    class="bp3-input-group"
                  >
                    <input
                      class="bp3-input"
                      type="text"
                      value="10"
                    />
                  </div>
                </div>
                <button
                  class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--withLeftIcon Button--variation Button--variation-secondary"
                  type="button"
                >
                  <span
                    class="bp3-icon bp3-icon-plus StyledProps--main"
                    icon="plus"
                  >
                    <svg
                      data-icon="plus"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <desc>
                        plus
                      </desc>
                      <path
                        d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
